src/domain/core/src/auth/dataset_action_authorizer.rs

```diff
@@ -93,22 +133,83 @@ pub enum DatasetActionUnauthorizedError {
     Access(AccessError),
 
     #[error(transparent)]
-    Internal(InternalError),
+    Internal(#[from] InternalError),
 }
 
 #[derive(Debug, Error)]
 #[error("User has no '{action}' permission in dataset '{dataset_ref}'")]
 pub struct DatasetActionNotEnoughPermissionsError {
     pub action: DatasetAction,
-    pub dataset_ref: DatasetRef,
+    pub dataset_ref: odf::DatasetRef,
 }
 
 ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
 #[derive(Debug)]
 pub struct ClassifyByAllowanceResponse {
-    pub authorized_handles: Vec<DatasetHandle>,
-    pub unauthorized_handles_with_errors: Vec<(DatasetHandle, DatasetActionUnauthorizedError)>,
+    pub authorized_handles: Vec<odf::DatasetHandle>,
+    pub unauthorized_handles_with_errors: Vec<(odf::DatasetHandle, DatasetActionUnauthorizedError)>,
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Extensions
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+pub trait DatasetActionAuthorizerExt: DatasetActionAuthorizer {
+    fn filtered_datasets_stream<'a>(
+        &'a self,
+        dataset_handles_stream: DatasetHandleStream<'a>,
+        action: DatasetAction,
+    ) -> DatasetHandleStream<'a>;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl<T> DatasetActionAuthorizerExt for T
+where
+    T: DatasetActionAuthorizer,
+    T: ?Sized,
+{
+    fn filtered_datasets_stream<'a>(
+        &'a self,
+        dataset_handles_stream: DatasetHandleStream<'a>,
+        action: DatasetAction,
+    ) -> DatasetHandleStream<'a> {
```

Comment:
Can this be used in GraphQL dependency routines?

Reply:
I reviewed some aspects of the code, and eventually added in:

- `ODataServiceContext::list_collections()`
- `Datasets::by_account_impl()`
- `Search::query()`